File: src/game_objects/apple.cpp
#include <stdlib.h>
#include "./apple.hpp"
#include "picosystem.hpp"

using namespace picosystem;

void apple_t::place(vec_t other) {
    do {
        pos.x = std::rand() % board->bounds->x;
        pos.y = std::rand() % board->bounds->y;
    } while(snake->collides(pos) && !collides(other));
    alive = true;
}

bool apple_t::collides(vec_t other) {
    return alive && pos.equals(other);
}

void apple_t::draw() {
    if (!alive) return;
    vec_t p = board->transform(pos);
    sprite(sprite_id, p.x, p.y);
}

File: src/game-objects/Apple.js
import Vector from '../Vector.js';

const {
    floor,
    random,
} = Math;

export default class Apple extends Vector {
    constructor({ board, spriteId, alive, pos = [] }) {
        super(...pos);
        Object.assign(this, { board, spriteId, alive });
    }
    place(...avoid) {
        do {
            this.board.bounds.forEach((bound, index) => {
                this[index] = floor(random() * bound);
            });
        } while (!avoid.reduce((avoided, vec) => avoided && !vec.collides(this), true));
        this.alive = true;
    }

    collides(other) {
        return this.alive && super(other); 
    }

    draw(tick) {
        if (!this.alive) return;
        const pos = this.board.transform(this);
        sprite(this.spriteId, pos.x, pos.y);
    }
}
